oauth/authcode.go

```diff
@@ -0,0 +1,183 @@
+package oauth
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os/exec"
+	"runtime"
+	"strings"
+	"time"
+
+	"context"
+
+	"github.com/danielgtaylor/openapi-cli-generator/cli"
+	"github.com/rs/zerolog/log"
+	"github.com/spf13/viper"
+	"golang.org/x/oauth2"
+	gcontext "gopkg.in/h2non/gentleman.v2/context"
+)
+
+// open opens the specified URL in the default browser regardless of OS.
+func open(url string) error {
+	var cmd string
+	var args []string
+
+	switch runtime.GOOS {
+	case "windows":
+		cmd = "cmd"
+		args = []string{"/c", "start"}
+	case "darwin":
+		cmd = "open"
+	default: // "linux", "freebsd", "openbsd", "netbsd"
+		cmd = "xdg-open"
+	}
+	args = append(args, url)
+	return exec.Command(cmd, args...).Start()
+}
+
+// authHandler is an HTTP handler that takes a channel and sends the `code`
+// query param when it gets a request.
+type authHandler struct {
+	c chan string
+}
+
+func (h authHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	h.c <- r.URL.Query().Get("code")
+	w.Header().Set("Content-Type", "text/html")
+	w.Write([]byte("<html><body><p>Login successful. Please return to the terminal. You may now close this window.</p></body></html>"))
+}
+
+// AuthorizationCodeTokenSource with PKCE as described in:
+// https://www.oauth.com/oauth2-servers/pkce/
+// This works by running a local HTTP server on port 8484 and then having the
```

Comment:
are we guaranteed that the local http server listening on 8484 is ours and not a malicious actor?

Reply:
Since we don't use `SO_REUSEADDR` when creating the socket for the HTTP server, if another (potentially malicious) actor is listening we will actually crash with a fatal error message rather than continue.